src/python/pants/goal/anonymous_telemetry.py

```diff
@@ -0,0 +1,185 @@
+    options_scope = "anonymous-telemetry"
+    help = (
+        "Options related to sending anonymous stats to the Pants project, to aid development. "
+        "Repos that opt into this telemetry will be able to access to their own stats."
+    )
+
+    @classmethod
+    def register_options(cls, register):
+        register(
+            "--enabled",
+            advanced=True,
+            type=bool,
+            help=(
+                f"Whether to send anonymous telemetry to the Pants project. Telemetry is sent "
+                f"asynchronously, with silent failure, and does not impact build times or "
+                f"outcomes. {_telemetry_docs_referral}."
+            ),
+        )
+        register(
+            "--repo-id",
+            advanced=True,
+            type=str,
+            help=(
+                f"An anonymized ID representing this repo.  For private repos, you likely want the "
+                f"ID to not be derived from, or algorithmically convertible to, anything "
+                f"identifying the repo.  For public repos the ID may be visible in that repo's "
+                f"config file, so anonymity of the repo is not guaranteed (although user anonymity "
+                f"is always guaranteed). {_telemetry_docs_referral}."
+            ),
+        )
+
+    @property
+    def enabled(self) -> bool:
+        return cast(bool, self.options.enabled)
+
+    @property
+    def repo_id(self) -> str | None:
+        return None if self.options.repo_id is None else cast(str, self.options.repo_id)
+
+
+class AnonymousTelemetryCallback(WorkunitsCallback):
+    def __init__(self, anonymous_telemetry: AnonymousTelemetry) -> None:
+        super().__init__()
+        self._anonymous_telemetry = anonymous_telemetry
+
+    @staticmethod
+    def validate_repo_id(repo_id: str) -> bool:
+        is_valid = re.match(r"^[a-zA-Z0-9-_]{30,60}$", repo_id) is not None
+        if not is_valid:
+            logger.error(
+                "The repo_id must be between 30 and 60 characters long, and consist of only "
+                "alphanumeric characters, dashes and underscores."
+            )
+        return is_valid
+
+    def __call__(
+        self,
+        *,
+        started_workunits: tuple[Workunit, ...],
+        completed_workunits: tuple[Workunit, ...],
+        finished: bool,
+        context: StreamingWorkunitContext,
+    ) -> None:
+        if not finished:
+            return
+
+        if self._anonymous_telemetry.options.is_default("enabled"):
+            logger.warning(
+                f"Please either set `enabled = true` in the [anonymous-telemetry] section of "
+                f"pants.toml to enable sending anonymous stats to the Pants project to aid "
+                f"development, or set `enabled = false` to disable it. No telemetry sent "
+                f"for this run. An explicit setting will get rid of this message. "
+                f"{_telemetry_docs_referral}."
```

Comment:
Consider adding `\n\n` to make it less of a big wall of text.

Reply:
I don't like line breaks in logs. Can use multiple log statements though.

Reply:
> I don't like line breaks in logs. 

How come?

> Can use multiple log statements though.

If that's the alternative, I'd rather stick with no newlines and a single log. It's really confusing with multiple logs if the message gets split up with unrelated logs being interspesed.

Reply:
But na, I think it looks best as is.